### PR TITLE
Feature/default methods

### PR DESCRIPTION
**Pull request by pavanupb**

### CODE REVIEW

**Comment by pavanupb** [comment]
@mbenz89 Could you please review the pull request and let me know for any changes

File: src/main/java/soot/jimple/toolkits/callgraph/OnFlyCallGraphBuilder.java [hunk 1]
```diff
@@ -206,28 +206,28 @@
   protected final HashSet<SootMethod> analyzedMethods = new HashSet<SootMethod>();
 
   // end type based reflection resolution
-  protected final LargeNumberedMap<Local, List<VirtualCallSite>> receiverToSites =
+  private final LargeNumberedMap<Local, List<VirtualCallSite>> receiverToSites =
```
Comment (mbenz89) [inline src/main/java/soot/jimple/toolkits/callgraph/OnFlyCallGraphBuilder.java:209]: Please make sure to make all of these fields `protected` as in `develop`

File: src/systemTest/java/soot/defaultInterfaceMethods/DefaultInterfaceTest.java [hunk 2]
```diff
@@ -0,0 +1,897 @@
...
+  public void interfaceSameSignatureTest() {
+    String testClassSig = "soot.defaultInterfaceMethods.InterfaceSameSignature";
+    String interfaceReadSig = "soot.defaultInterfaceMethods.Read";
+    String interfaceWriteSig = "soot.defaultInterfaceMethods.Write";
+
+    final SootMethod target =
+        prepareTarget(
+            methodSigFromComponents(testClassSig, "void", "main"),
+            testClassSig,
+            "soot.defaultInterfaceMethods.Read",
```
Comment (mbenz89) [inline src/systemTest/java/soot/defaultInterfaceMethods/DefaultInterfaceTest.java:118]: Replace by variables

File: src/systemTest/java/soot/defaultInterfaceMethods/DefaultInterfaceTest.java [hunk 3]
```diff
@@ -0,0 +1,897 @@
...
+    boolean edgeMainPrintToWritePrint = checkInEdges(writeInterfacePrint, mainPrintMethod);
+    boolean edgeMainMethodToPrint = checkInEdges(mainPrintMethod, target);
+    boolean edgeMainMethodToReadMethod = checkInEdges(readInterfaceRead, target);
+    boolean edgeMainMethodToWriteMethod = checkInEdges(writeInterfaceWrite, target);
+
+    /* Edges should not be present */
+    boolean edgeMainMethodToReadPrint = checkInEdges(readInterfacePrint, target);
+    boolean edgeMainMethodToWritePrint = checkInEdges(writeInterfacePrint, target);
+
+    final ReachableMethods reachableMethods = Scene.v().getReachableMethods();
```
Comment (mbenz89) [inline src/systemTest/java/soot/defaultInterfaceMethods/DefaultInterfaceTest.java:201]: Could you please add your test case `CallGraphTest` here as separet @Test method? Move the target files to the `soot.defaultInterfaceMethods` directory in test-targets and then copy paste the head of one of this test methods with the correct target class: `  String testClassSig = "soot.defaultInterfaceMethods.PavansSuperCoolTest";`

File: eclipse/ca.mcgill.sable.soot/src/ca/mcgill/sable/soot/ui/PhaseOptionsDialog.java [hunk 4]
```diff
@@ -9119,18 +9122,6 @@ private Composite Output_OptionsCreate(Composite parent) {
 						"1.9 9",
 						"\nForce Java 1.9 as output version. (Experimental)",
 						false),
-				new OptionData("Java 1.10",
```
Comment (mbenz89) [inline eclipse/ca.mcgill.sable.soot/src/ca/mcgill/sable/soot/ui/PhaseOptionsDialog.java:9122]: Why are these removed? Werent they added by the default branch?

File: src/main/generated/options/soot/AntTask.java [hunk 5]
```diff
@@ -370,12 +371,6 @@ public void setjava_version(String arg) {
                 || arg.equals( "8" )
                 || arg.equals( "1.9" )
                 || arg.equals( "9" )
-                || arg.equals( "1.10" )
```
Comment (mbenz89) [inline src/main/generated/options/soot/AntTask.java:373]: Same here. Should be added??

File: src/main/generated/options/soot/options/Options.java [hunk 6]
```diff
@@ -100,12 +100,6 @@ public static Options v() {
     public static final int java_version_8 = 9;
     public static final int java_version_1_9 = 10;
     public static final int java_version_9 = 10;
-    public static final int java_version_1_10 = 11;
```
Comment (mbenz89) [inline src/main/generated/options/soot/options/Options.java:103]: Same here. Should be added??

File: eclipse/ca.mcgill.sable.soot/src/ca/mcgill/sable/soot/ui/PhaseOptionsDialog.java (hunk 4 above)
Comment (mbenz89) [reply eclipse/ca.mcgill.sable.soot/src/ca/mcgill/sable/soot/ui/PhaseOptionsDialog.java:9122]: MIght it be that they get re-added when you call ``mvn clean verify`?

Comment (pavanupb) [reply eclipse/ca.mcgill.sable.soot/src/ca/mcgill/sable/soot/ui/PhaseOptionsDialog.java:9122]: Ran mvn clean verify. But did not get any changes. Will check this

File: src/systemTest/java/soot/defaultInterfaceMethods/DefaultInterfaceTest.java (hunk 3 above)
Comment (pavanupb) [reply src/systemTest/java/soot/defaultInterfaceMethods/DefaultInterfaceTest.java]: Okay

File: eclipse/ca.mcgill.sable.soot/src/ca/mcgill/sable/soot/ui/PhaseOptionsDialog.java (hunk 4 above)
Comment (pavanupb) [reply eclipse/ca.mcgill.sable.soot/src/ca/mcgill/sable/soot/ui/PhaseOptionsDialog.java:9122]: I see this has been removed in the develop branch https://github.com/soot-oss/soot/blob/e233653f2a1bd823d20b490ff75ccc50033ff421/eclipse/ca.mcgill.sable.soot/src/ca/mcgill/sable/soot/ui/PhaseOptionsDialog.java#L9122

File: src/main/generated/options/soot/AntTask.java (hunk 5 above)
Comment (pavanupb) [reply src/main/generated/options/soot/AntTask.java:373]: This is also removed in the develop branch https://github.com/soot-oss/soot/blob/e233653f2a1bd823d20b490ff75ccc50033ff421/src/main/generated/options/soot/AntTask.java#L373

File: src/main/generated/options/soot/options/Options.java (hunk 6 above)
Comment (pavanupb) [reply src/main/generated/options/soot/options/Options.java:103]: This is also not present in develop https://github.com/soot-oss/soot/blob/e233653f2a1bd823d20b490ff75ccc50033ff421/src/main/generated/options/soot/options/Options.java#L102

File: eclipse/ca.mcgill.sable.soot/src/ca/mcgill/sable/soot/ui/PhaseOptionsDialog.java (hunk 4 above)
Comment (mbenz89) [reply eclipse/ca.mcgill.sable.soot/src/ca/mcgill/sable/soot/ui/PhaseOptionsDialog.java:9122]: No! This was added in the default branch! Thats why it is not in develop.

File: src/main/generated/options/soot/AntTask.java (hunk 5 above)
Comment (mbenz89) [reply src/main/generated/options/soot/AntTask.java:373]: Look at the default branch please. Also double check that you have not removed other things i added there please.

File: eclipse/ca.mcgill.sable.soot/src/ca/mcgill/sable/soot/ui/PhaseOptionsDialog.java (hunk 4 above)
Comment (pavanupb) [reply eclipse/ca.mcgill.sable.soot/src/ca/mcgill/sable/soot/ui/PhaseOptionsDialog.java:9122]: Okay. Will add them.

File: src/main/generated/options/soot/AntTask.java (hunk 5 above)
Comment (pavanupb) [reply src/main/generated/options/soot/AntTask.java:373]: Okay

File: src/main/generated/options/soot/AntTask.java [hunk 7]
```diff
@@ -223,10 +228,6 @@ public Path createprocess_dir() {
             return process_dir.createPath();
         }
   
-        public void setderive_java_version(boolean arg) {
```
Comment (pavanupb) [inline src/main/generated/options/soot/AntTask.java:226]: @mbenz89 Do we need this method?

Comment (mbenz89) [reply src/main/generated/options/soot/AntTask.java:226]: This is auto generated. You will just have to fix the sot_options.xml file and then some files will be generated. Try it :)

Comment (pavanupb) [reply src/main/generated/options/soot/AntTask.java:226]: But I dont see any difference from our default branch to develop branch

File: src/systemTest/java/soot/defaultInterfaceMethods/DefaultInterfaceTest.java (hunk 2 above)
Comment (pavanupb) [reply src/systemTest/java/soot/defaultInterfaceMethods/DefaultInterfaceTest.java]: Should I replace for all the test methods?

Comment (mbenz89) [reply src/systemTest/java/soot/defaultInterfaceMethods/DefaultInterfaceTest.java]: If there are already varaibles existing for these strings then yes! We do not want to have duplicate to be able to more easily change things later.
